src/analytics.js

```diff
@@ -105,3 +110,107 @@ export function insertAnalyticsElement(
   parentElement.appendChild(analyticsElem);
   return analyticsElem;
 }
+
+/**
+ * A class that is used by extension to use Amp analytics features.
+ * The class create/remove sandbox amp-analytics element at right times.
+ */
+export class SandboxAnalyticsAdapter {
+  /**
+   * @param {!AmpElement} parent
+   * @param {function():!Promise<!JSONType>} createConfigFunc
+   */
+  constructor(parent, createConfigFunc) {
+
+    /** @private {!AmpElement} */
+    this.parent_ = parent;
+
+    /** @private {!function():!Promise<!JSONType>} */
+    this.createConfigFunc_ = createConfigFunc;
+
+    /** {?Element} */
+    this.analyticsElement = null;
+
+    /** @private {?Promise<!JSONType>} */
+    this.getConfigPromise_ = null;
+
+    /** @private {?function()} */
+    this.analyticsCreatedResolver_ = null;
+
+    this.listenToLoadStart_();
+  }
+
+  /**
+   * Return promise that resolve on analytics element has laid out and is ready
+   * Please note: analytics element may be removed during unlayout, thus this
+   * function always return ready promise that is tied to current round of layout
+   * @return {!Promise}
+   */
+  onAnalyticsReady() {
+    let analyticsElementCreatedPromise;
+    if (this.analyticsElement) {
+      analyticsElementCreatedPromise = Promise.resolve();
+    } else {
+      analyticsElementCreatedPromise = new Promise(resolve => {
+        this.analyticsCreatedResolver_ = resolve;
+      });
+    }
+    return analyticsElementCreatedPromise.then(() => {
+      return this.analyticsElement.signals().whenSignal('amp-analytics-ready');
+    });
+  }
+
+  /**
+   * Private function that get config and insert analytics element
+   */
+  createSandboxAnalyticsElement_() {
+    if (this.analyticsElement || this.getConfigPromise_) {
+      return;
+    }
+    this.getConfigPromise_ = this.createConfigFunc_();
+    this.getConfigPromise_.then(config => {
+      if (!this.getConfigPromise_) {
```

Comment:
If we can unload then reload, there's a chance that a new promise is here. Check that they are the same instance.